vault/eventbus/bus_test.go

```diff
@@ -180,3 +184,177 @@ func TestBus2Subscriptions(t *testing.T) {
 		t.Error("Timeout waiting for event2")
 	}
 }
+
+// TestBusSubscriptionsCancel verifies that canceled subscriptions are cleaned up.
+func TestBusSubscriptionsCancel(t *testing.T) {
+	subscriptions.Store(0)
+	bus, err := NewEventBus(nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ctx := context.Background()
+	bus.Start()
+
+	// create and cancel a bunch of subscriptions
+	create := 100
+	cancel := 50
+
+	eventType := logical.EventType("someType")
+
+	var channels []<-chan *logical.EventReceived
+	var cancels []context.CancelFunc
+	stopped := atomic.Int32{}
+	canceled := atomic.Int32{}
+	stopped.Store(-1)
+
+	received := atomic.Int32{}
+
+	for i := 0; i < create; i++ {
+		ch, cancelFunc, err := bus.Subscribe(ctx, namespace.RootNamespace, eventType)
+		if err != nil {
+			t.Fatal(err)
+		}
+		t.Cleanup(cancelFunc)
+		channels = append(channels, ch)
+		cancels = append(cancels, cancelFunc)
+
+		go func(i int32) {
+			<-ch
+			received.Add(1)
+			for i < int32(cancel) {
+				<-ch
+				received.Add(1)
+			}
+			canceled.Add(1)
+			cancelFunc() // cancel explicitly to unsubscribe
+		}(int32(i))
+	}
+
+	// check that all channels receive a message
+	event, err := logical.NewEvent()
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = bus.SendInternal(ctx, namespace.RootNamespace, nil, eventType, event)
+	if err != nil {
+		t.Error(err)
+	}
+	waitFor(t, 1*time.Second, func() bool { return received.Load() == int32(create) })
+
+	if canceled.Load() != int32(cancel) {
```

Comment:
Should this be a `waitFor`? I don't think there's any waiting before here that guarantees all the goroutines will have got far enough.

Reply:
Good catch!